docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,463 @@
+Most at risk is service location:
+information regarding the host and port where an EdgeX service is available.
+To a lesser extent,
+Consul holds general service configuration information
+and is the fallback location for the EdgeX secret store if Vault is not used.
+Consul also provides for service health checks.
+
+The elements required to secure Consul are:
+
+- For neighbor discovery: An **encryption key** used by Serf;
+  this key must be provided the first time that Consul is started.
+  This key prevents rogue Consul instances from joining the cluster.
+  This key must also be provided to any new Consul agents that are added to the system.
+  This key can be a 32-byte derived key or a random key.
+- For TLS: a **CA certificate**, **TLS certificate (chain)**, and **TLS private key**.
+
+For integration with Vault,
+some initial secrets must be created:
+
+- A **management token** that can be passed to Vault's Consul secrets engine,
+  to enable a Vault API to be utilized to create additional per-microservice Consul tokens.
+- An **ACL token** that allows Vault to read/write a portion of Consul's key space as the backing store for Vault.
+  (Intent is to switch to the Vault raft storage backend to avoid Consul dependency.)
+
+#### EdgeX secret store
+
+The EdgeX secret store is based on Hashicorp Vault.
+Upon initialization, Vault generates a **master key**,
+which is used to encrypt the secrets database.
+The master key is used to unseal Vault secrets,
+and to regenerate the Vault root token.
+
+#### EdgeX public key infrastructure
+
+A cluster-wide PKI is needed
+for TLS-based confidentiality and integrity protection
+of EdgeX microservice communication.
+
+EdgeX will use Hashicorp Vault to secure the EdgeX PKI,
+through the use of the Vault PKI secrets engine.
+Vault will be configured with a root CA at initialization time,
+and a Vault-based sub-CA for dynamic generation of TLS leaf certificates.
+The root CA will be restricted to use only by the Vault root token.
+
+EdgeX microservices that are based on third-party containers
+require special support unless they can talk natively to Vault for their secrets.
+Certain tools, such as those mentioned in the
+"Creation and Distribution of Secrets" ADR
+(`envconsul`, `consul-template`, and others)
+can be used to facilitiate third-party container integration.
+Additionally, certain EdgeX services are considered bootstrapping services,
+and require special support
+because these components are involved in brining up the EdgeX secret store itself.
+These services are:
+
+* **Consul**: Requires TLS certificate set by configuration file or command line, with a TLS certificate injected into the container.
+
+* **Vault**: As Vault's database is encrypted, Vault cannot natively bootstrap its own TLS certificate.  Requires TLS certificate to be injected into container and its location set in a configuration file.
+
+* **PostgreSQL**: Requires TLS certificate to be injected into '$PGDATA'.
```

Comment:
what is this $PGDATA?  can you elaborate a little bit more

Reply:
Whoops!  https://github.com/edgexfoundry/developer-scripts/issues/294.  PGDATA is where postgres writes its data files.  You have (perhaps unwittingly) discovered that we aren't mounting a docker volume there.

Reply:
documentation fixed (bug filed)

Reply:
good to know